extern/storage-sealing/terminate_batch.go

```diff
@@ -0,0 +1,264 @@
+	"bytes"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-bitfield"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/dline"
+	miner2 "github.com/filecoin-project/specs-actors/v2/actors/builtin/miner"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/actors/builtin/miner"
+)
+
+var (
+	// TODO: config
+
+	TerminateBatchMax  uint64 = 100 // adjust based on real-world gas numbers, actors limit at 10k
+	TerminateBatchMin  uint64 = 1
+	TerminateBatchWait        = 5 * time.Minute
+)
+
+type TerminateBatcherApi interface {
+	StateSectorPartition(ctx context.Context, maddr address.Address, sectorNumber abi.SectorNumber, tok TipSetToken) (*SectorLocation, error)
+	SendMsg(ctx context.Context, from, to address.Address, method abi.MethodNum, value, maxFee abi.TokenAmount, params []byte) (cid.Cid, error)
+	StateMinerInfo(context.Context, address.Address, TipSetToken) (miner.MinerInfo, error)
+	StateMinerProvingDeadline(context.Context, address.Address, TipSetToken) (*dline.Info, error)
+}
+
+type TerminateBatcher struct {
+	api     TerminateBatcherApi
+	maddr   address.Address
+	mctx    context.Context
+	addrSel AddrSel
+	feeCfg  FeeConfig
+
+	todo map[SectorLocation]*bitfield.BitField // MinerSectorLocation -> BitField
+
+	waiting map[SectorLocation][]chan cid.Cid
+
+	notify, stop, stopped chan struct{}
+	force                 chan chan *cid.Cid
+	lk                    sync.Mutex
+}
+
+func NewTerminationBatcher(mctx context.Context, maddr address.Address, api TerminateBatcherApi, addrSel AddrSel, feeCfg FeeConfig) *TerminateBatcher {
+	b := &TerminateBatcher{
+		api:     api,
+		maddr:   maddr,
+		mctx:    mctx,
+		addrSel: addrSel,
+		feeCfg:  feeCfg,
+
+		todo:    map[SectorLocation]*bitfield.BitField{},
+		waiting: map[SectorLocation][]chan cid.Cid{},
+
+		notify:  make(chan struct{}, 1),
+		force:   make(chan chan *cid.Cid),
+		stop:    make(chan struct{}),
+		stopped: make(chan struct{}),
+	}
+
+	go b.run()
+
+	return b
+}
+
+func (b *TerminateBatcher) run() {
+	var forceRes chan *cid.Cid
+	var lastMsg *cid.Cid
+
+	for {
+		if forceRes != nil {
+			forceRes <- lastMsg
+			forceRes = nil
+		}
+		lastMsg = nil
+
+		var notif, after bool
+		select {
+		case <-b.stop:
+			close(b.stopped)
+			return
+		case <-b.notify:
+			notif = true // send above max
+		case <-time.After(TerminateBatchWait):
+			after = true // send above min
+		case fr := <-b.force: // user triggered
+			forceRes = fr
+		}
+
+		dl, err := b.api.StateMinerProvingDeadline(b.mctx, b.maddr, nil)
+		if err != nil {
+			log.Errorw("TerminateBatcher: getting proving deadline info failed", "error", err)
+			continue
+		}
+
+		b.lk.Lock()
+		params := miner2.TerminateSectorsParams{}
+
+		var total uint64
+		for loc, sectors := range b.todo {
```

Comment:
We also need to limit to at most `miner.DeclarationsMax` (currently 3000). The current max will cover that, but it would be nice to have some explicit upper bounds here.